Automattic/jetpack
# VideoPress: Route video search using query parameters

Relates to #26720.

#### Changes proposed in this Pull Request:

* Refactor the URL parameter hook to support multiple parameters
* Add support to the `q` parameter to route video search, so:
   * when some query is typed on the search box, the URL is updated
   * when someone lands on the VideoPress library with a `q` parameter, the search is triggered
   * when an empty search is made, the parameter is removed from the URL
   * when accessing a page with a `q=` parameter, the search field is filled with the given string

#### Other information:


#### Jetpack product discussion



#### Does this pull request change what data or activity we track or use?



#### Testing instructions:






* Make sure you test on a site with multiple videos, so you can try multiple searches and can have search results with multiple pages
* Go to `Jetpack > VideoPress`
* Type in some search term on the search box:

<img width="700" alt="Screen Shot 2023-01-04 at 11 56 29" src="https://user-images.githubusercontent.com/6760046/210582971-258c0ede-205e-4e80-bf37-2db6aeb869c2.png">

* Notice that the term that you searched for is now present on the query string portion of the URL, on the `q` parameter:

<img width="269" alt="Screen Shot 2023-01-04 at 11 57 32" src="https://user-images.githubusercontent.com/6760046/210583193-d8c9bb6e-c0b4-44d7-a4a1-692256dde471.png">

* Try cleaning the search box (using the "x" button or just removing the text using the keyboard):
   * Notice that the `q` parameter is removed from the URL

<img width="700" alt="Screen Shot 2023-01-04 at 12 00 21" src="https://user-images.githubusercontent.com/6760046/210583754-45c6487c-d476-401f-8a0a-7221a9f09890.png">

* Search for some term, copy the resulting URL (with the `q` filter included), and then open it using a new tab:
   * Notice that the search term is applied to the video results
   * Notice that the search term is present on the search field

<img width="700" alt="Screen Shot 2023-01-04 at 12 10 06" src="https://user-images.githubusercontent.com/6760046/210586211-677cd942-f49c-4d6e-b0bf-66b804136662.png">

* Search for some term that may result on multiple pages of videos:
   * Notice that the pagination still works as expected, keeping the page number on the URL
   * Notice that the search term is present on the URL for every page of video results

<img width="700" alt="Screen Shot 2023-01-04 at 12 13 44" src="https://user-images.githubusercontent.com/6760046/210586925-40b6dc06-c893-4bee-905f-10f85e9e3f23.png">

* Try doing a search on a page other than the first
   * Notice that the page number is reset so the user can start seeing the searched videos from page 1
* Try changing the search term on the URL
   * Notice that the videos get filtered by the new search term

Known issues that I am planning to tackle on a followup PR:

* when accessing a page with a `q=` parameter, the videos get filtered but in some specific cases one video from the original data set is showed within the filtered ones; (I thought this could be related to #27848 but it's not, and it's also happens on very specific scenarios); I will write a followup issue for this, trying to find a reliable way to replicate it
* when accessing a page with a `q=` parameter, it may be possible to see a flick of the original video list from the `initialState` var, that was loaded before the list of searched videos got showed
